pkg/controller/agent/pod.go

```diff
@@ -50,63 +63,199 @@ var (
+	vols := []volume.VolumeLike{volume.NewSecretVolume(
+		ConfigSecretName(params.Agent.Name),
+		FleetSetupVolumeName,
+		path.Join(FleetSetupMountPath, FleetSetupFileName),
+		FleetSetupFileName,
+		0440,
+	)}
+
+	var esAssociation commonv1.Association
+	//nolint:nestif
+	if params.Agent.Spec.FleetServerEnabled {
+		// ECK creates CA and a certificate for Fleet Server to use. This volume contains those.
+		vols = append(vols, volume.NewSecretVolumeWithMountPath(
+			fleetCerts.Name,
+			FleetCertsVolumeName,
+			FleetCertsMountPath,
+		))
+
+		builder = builder.WithPorts([]corev1.ContainerPort{{Name: params.Agent.Spec.HTTP.Protocol(), ContainerPort: FleetServerPort, Protocol: corev1.ProtocolTCP}})
+
+		// Beats managed by the Elastic Agent don't trust the Elasticsearch CA that Elastic Agent itself is configured
+		// to trust. There is currently no way to configure those Beats to trust a particular CA. The intended way to handle
+		// it is to allow Fleet to provide Beat output settings, but due to https://github.com/elastic/kibana/issues/102794
+		// this is not supported outside of UI. To workaround this limitation the Agent is going to update Pod-wide CA store
+		// before starting Elastic Agent.
+		esAssociation = association.GetAssociationOfType(params.Agent.GetAssociations(), commonv1.ElasticsearchAssociationType)
+	} else {
+		// See the long comment above. As the reference chain is: Elastic Agent ---> Fleet Server ---> Elasticsearch,
+		// we need first to identify the Fleet Server and then identify its reference to Elasticsearch.
+		fs, err := getAssociatedFleetServer(params)
+		if err != nil {
+			return nil, err
+		}
+
+		if fs != nil {
+			esAssociation = association.GetAssociationOfType(fs.GetAssociations(), commonv1.ElasticsearchAssociationType)
+			if esAssociation != nil {
+				if params.Agent.Namespace != esAssociation.Associated().GetNamespace() {
+					return nil, fmt.Errorf(
+						"agent namespace %s is different than referenced Elasticsearch namespace %s, this is not supported yet",
+						params.Agent.Namespace,
+						esAssociation.Associated().GetNamespace(),
+					)
+				}
+
+				caSecretName := esAssociation.AssociationConf().GetCASecretName()
+				vols = append(vols, volume.NewSecretVolumeWithMountPath(
+					caSecretName,
+					fmt.Sprintf("%s-certs", esAssociation.AssociationType()),
+					certificatesDir(esAssociation),
+				))
+
+				// Because of the reference chain (Elastic Agent ---> Fleet Server ---> Elasticsearch), we are going to get
+				// notified when CA of Elasticsearch changes as Fleet Server resource will get updated as well. But what we
+				// also need to do is to roll Elastic Agent Pods to pick up the update CA. To do be able to do that, we are
+				// adding Fleet Server associations (which includes Elasticsearch) to config hash attached to Elastic Agent
+				// Pods.
+				if err := commonassociation.WriteAssocsToConfigHash(params.Client, fs.GetAssociations(), configHash); err != nil {
```

Comment:
ugh 😬

Reply:
Could you elaborate? I'm not sure how I can address this comment.